RoundsWithFriends/Patches/PlayerManager.cs

```diff
@@ -38,4 +47,143 @@ static bool Prefix(PlayerManager __instance, Player asker, ref Player __result)
+            }
+
+            return codes.AsEnumerable();
+        }
+    }
+
+    [HarmonyPatch(typeof(PlayerManager), "MovePlayers")]
+    class PlayerManager_Patch_MovePlayers
+    {
+        static bool Prefix(PlayerManager __instance, SpawnPoint[] spawnPoints)
+        {
+            __instance.StartCoroutine(WaitForMapToLoad(__instance, spawnPoints));
+            return false;
+        }
+        private static LayerMask groundMask = (LayerMask) LayerMask.GetMask(new string[] { "Default", "IgnorePlayer" });
+        static bool MapHasValidGround(Map map)
+        {
+            if (!(bool)map.GetFieldValue("hasCalledReady")) { return false; }
+
+            foreach (Transform transform in map.gameObject.transform)
+            {
+                RaycastHit2D raycastHit2D = Physics2D.Raycast(transform.position + 100f*Vector3.up, Vector2.down, 101f, groundMask);
+                if (raycastHit2D.transform && raycastHit2D.distance > 0.1f)
+                {
+                    Vector2 screenPoint = MainCam.instance.transform.GetComponent<Camera>().FixedWorldToScreenPoint(raycastHit2D.point);
+                    screenPoint.x /= FixedScreen.fixedWidth;
+                    screenPoint.y /= Screen.height;
+                    if (screenPoint.x >= 0f && screenPoint.x <= 1f && screenPoint.y >= 0f && screenPoint.y <= 1f)
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        static IEnumerator WaitForMapToLoad(PlayerManager __instance, SpawnPoint[] spawnPoints)
+        {
+            // wait until the map has solid ground
+            yield return new WaitUntil(() => MapHasValidGround(MapManager.instance.currentMap?.Map));
+            // 10 extra frames to make the game happy
+            for (int _ = 0; _ < 10; _++)
+            {
+                yield return null;
+            }
+
+            Dictionary<Player, Vector2> spawnDictionary = GeneralizedSpawnPositions.GetSpawnDictionary(__instance.players, spawnPoints);
+
+            for (int i = 0; i < __instance.players.Count; i++)
+            {
+
+                __instance.StartCoroutine((IEnumerator) typeof(PlayerManager).InvokeMember("Move",
+                    BindingFlags.Instance | BindingFlags.InvokeMethod |
+                    BindingFlags.NonPublic, null, __instance, new object[] { __instance.players[i].data.playerVel, (Vector3) spawnDictionary[__instance.players[i]] }));
+
+                // I have no idea why this is in the original method like this but I'm afraid to change it
+                int j;
+                for (j = i; j >= __instance.soundCharacterSpawn.Length; j -= __instance.soundCharacterSpawn.Length)
+                {
+                }
+
+                SoundManager.Instance.Play(__instance.soundCharacterSpawn[j], __instance.players[i].transform);
```

Comment:
This retarded piece of code is assumably trying to say `int j = i % __instance.soundCharacterSpawn.Length;`. Could just be a weird transformation made by the compiler.